token-lending/program/src/processor.rs

```diff
@@ -0,0 +1,1636 @@
+    }
+    if &borrow_reserve.liquidity.supply_pubkey == destination_liquidity_info.key {
+        msg!("Cannot use borrow reserve liquidity supply as destination account input");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    // TODO: handle case when neither reserve is the quote currency
+    if borrow_reserve.dex_market.is_none() && deposit_reserve.dex_market.is_none() {
+        msg!("One reserve must have a dex market");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if let COption::Some(dex_market_pubkey) = borrow_reserve.dex_market {
+        if &dex_market_pubkey != dex_market_info.key {
+            msg!("Invalid dex market account input");
+            return Err(LendingError::InvalidAccountInput.into());
+        }
+    }
+    if let COption::Some(dex_market_pubkey) = deposit_reserve.dex_market {
+        if &dex_market_pubkey != dex_market_info.key {
+            msg!("Invalid dex market account input");
+            return Err(LendingError::InvalidAccountInput.into());
+        }
+    }
+
+    let mut obligation = Obligation::unpack(&obligation_info.data.borrow())?;
+    if obligation_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &obligation.borrow_reserve != borrow_reserve_info.key {
+        msg!("Borrow reserve input doesn't match existing obligation borrow reserve");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &obligation.collateral_reserve != deposit_reserve_info.key {
+        msg!("Collateral reserve input doesn't match existing obligation collateral reserve");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    unpack_mint(&obligation_token_mint_info.data.borrow())?;
+    if &obligation.token_mint != obligation_token_mint_info.key {
+        msg!("Obligation token mint input doesn't match existing obligation token mint");
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    let obligation_token_output = Token::unpack(&obligation_token_output_info.data.borrow())?;
+    if obligation_token_output_info.owner != token_program_id.key {
+        return Err(LendingError::InvalidTokenOwner.into());
+    }
+    if &obligation_token_output.mint != obligation_token_mint_info.key {
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    assert_last_update_slot(&borrow_reserve, clock.slot)?;
+    assert_last_update_slot(&deposit_reserve, clock.slot)?;
+    obligation.accrue_interest(borrow_reserve.cumulative_borrow_rate_wads)?;
+
+    let trade_simulator = TradeSimulator::new(
+        dex_market_info,
+        dex_market_orders_info,
+        memory,
+        &lending_market.quote_token_mint,
+        &borrow_reserve.liquidity.mint_pubkey,
+        &deposit_reserve.liquidity.mint_pubkey,
+    )?;
+
+    let loan = deposit_reserve.create_loan(
+        collateral_amount,
+        loan_amount,
```

Comment:
Do we want to cap the leverage amount here? Or we will leave for later? Sorry if I missed if it was added somewhere in this PR

Reply:
I can cap it. would be a reserve parameter that is passed in when the reserve is created. will make the change now.

Reply:
pulled this work out into its own branch